### PR TITLE
Fixing "Error handling difficult and not reliable #38"

### PR DESCRIPTION
**Pull request by noneymous**
Here comes the implementation of my suggestion to fix https://github.com/Ullaakut/nmap/issues/38. Please have a look and see how this works for you. I tested it against some scan targets and am still running it against even more. **You are probably using different Nmap arguments than me, so I'm curious about your experiences.** 

Result processing would be like:
1) Check if error is returned (includes package errors and nmap errors). If error is returned, something went most probably wrong and you can abort. Warning's might give you more details/flexibility. Most returned errors are defined as error constants in errors.go. Have a look there to see what issues might come back to handle them individually
2) OptionA: Check warnings for things that might be an issue for you and abort or proceed
OptionB: Ignore warnings at all and just work with the returned XML. It already contains a lot of details

Here is sample usage code:

```go
// Prepare Nmap scan
proc, err := nmap.NewScanner(options...)
if err != nil {
    fmt.Println("Could not initialize Nmap")
    return
}

// Execute Nmap scan
result, warnings, err := proc.Run()

// Check for nmap errors and optionally handle them individually
if err != nil {
	switch err {
		case nmap.ErrScanTimeout:
        	fmt.Println("ERROR: Nmap scan ran into timeout")
    	case nmap.ErrParseOutput:
        	fmt.Printf("ERROR: Nmap output could not be parsed: %w\n", warnings[len(warnings)-1])
    	case nmap.ErrExcludeList:
        	fmt.Println("ERROR: Nmap could not resolve host(s) on exclude list")
    	default:
        	fmt.Printf("ERROR: Nmap scan failed with unexpected error: %w\n", err)
	}
	return
}

// Check for nmap warnings that might be relevant for you
for _, warning := range warnings {
    switch {
 		case strings.Contains(warning, "Failed to resolve"):
        	fmt.Println("A target could not be resolved.") 
    	case strings.Contains(warning, "No targets were specified")
			fmt.Println("The only target is blacklisted.")
    	case strings.Contains(warning, "TTVAR has grown")
        	fmt.Println("Just some TTVAR change notification.")
    	default:
        	fmt.Println(fmt.Sprintf("Other warning: %s", warning))
    }
}

// Process result
fmt.Println(result)
```

### CODE REVIEW

File: nmap.go [hunk 1]
```diff
@@ -56,7 +54,7 @@ func NewScanner(options ...func(*Scanner)) (*Scanner, error) {
 }
 
 // Run runs nmap synchronously and returns the result of the scan.
-func (s *Scanner) Run() (*Run, error) {
+func (s *Scanner) Run() (*Run, []string, error) {
```
Comment (Ullaakut) [inline nmap.go]: WDYT of naming the returns here, to make it easier to understand what the function returns by just looking at the prototype?

```suggestion
func (s *Scanner) Run() (result *Run, warnings []string, err error) {
```

File: nmap.go [hunk 2]
```diff
@@ -80,39 +80,53 @@ func (s *Scanner) Run() (*Run, error) {
...
 
-		return nil, ErrScanTimeout
+		return nil, []string{}, ErrScanTimeout
 	case <-done:
-		// Scan finished before timeout.
-		var nmapErrors []string
+
+		// Process nmap stderr output containing none-critical errors and warnings
+		// Everyone needs to check whether one or some of these warnings is a hard issue in their use case
+		var warnings []string
```
Comment (Ullaakut) [inline nmap.go]: If the returns are named, we can remove this line as well

File: nmap.go [hunk 3]
```diff
@@ -80,39 +80,53 @@ func (s *Scanner) Run() (*Run, error) {
...
+	// Wait for nmap process or timeout
 	select {
 	case <-s.ctx.Done():
+
 		// Context was done before the scan was finished.
 		// The process is killed and a timeout error is returned.
 		_ = cmd.Process.Kill()
 
-		return nil, ErrScanTimeout
+		return nil, []string{}, ErrScanTimeout
```
Comment (Ullaakut) [inline nmap.go]: If the returns are named, we can directly return `warnings`, which will be an empty string slice

File: nmap_test.go [hunk 4]
```diff
@@ -181,7 +181,7 @@ func TestRun(t *testing.T) {
 				panic(err) // this is never supposed to err, as we are testing run and not new.
 			}
 
-			result, err := s.Run()
+			result, _, err := s.Run()
```
Comment (Ullaakut) [inline nmap_test.go]: We should test that the warnings work as expected though. If you don't know how to do this, let me know and I'll gladly write the unit tests 👍

**Comment by noneymous** [comment]
Good idea, I introduced named return values and it's already commited. 

Just the warning unit test outstanding. Maybe tomorrow.

**Comment by noneymous** [comment]
Sorry, I'm in the wrong environment to get the tests running. I feel this would take me too much time but might be a simple thing for you...

File: nmap_test.go (hunk 4 above)
Comment (TheSecEng) [reply nmap_test.go]: Agreed

**Comment by TheSecEng** [comment]
Awaiting Unit Tests

**Comment by Ullaakut** [comment]
I'll test it this weekend :) It's on my calendar now:

Screenshot 2019-11-22 at 10 20 42 AM

**Comment by noneymous** [comment]
Thank you guys

**Comment by noneymous** [comment]
What's the plan, when do you intend to merge?

**Comment by TheSecEng** [comment]
@noneymous we will merge once tests have been completed. We thank you for your patience. Both myself and @Ullaakut have busy schedules at times.

**Comment by Ullaakut** [comment]
I'll wait for @TheSecEng to test it as well and once it's fine for him, we can merge the PR :)

**Comment by TheSecEng** [comment]
Cleared.

**Pull request merged**